ER-785 Performance testing remedial work

chrisjensenuk opened the pull request:
Employer Web was quite chatty so these changes should be quite noticeable.

We were checking whether a collection exists before every command sent to Cosmos DB. I’ve removed this check and have created a `MongoDbCollectionChecker` that runs once during the start of the Jobs to ensure all collections are present.

Fluent validators were not registered as singletons in Employer.Web so multiple times per request these were recreated which meant multiple calls to retrieve reference data on every page load.

Rebuild of the dashboard (which is done in-process every time a draft vacancy is updated) would call ReferenceData for every vacancy an employer has to retrieve training programmes. I’ve now cached training programmes.

I have added caching for all ReferenceData providers. All are cached aside until midnight except Apprenticeship Programmes which expire at 4:15am as our population job runs at 4am.

`MongoDbVacancyReviewRepository.GetActiveAsync` is used when generating the QA dashboard.  I've now added a projection so it only returns a subset of data. Hopefully this should help with the reported error `Error executing Mongo Command for method GetActiveAsync Reason: Command getMore failed: Unable to find cursor with id: 7.. Retrying in 1 secs...attempt: 1`

In `MongoDbCollectionBase` I've added a helper function `LogMongoCommand` which is commented out but can be used during debugging to log out what json gets sent to Cosmos.

sachpatel commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/Mongo/MongoDbCollectionChecker.cs:
```diff
@@ -0,0 +1,69 @@
+
+namespace Esfa.Recruit.Vacancies.Client.Infrastructure.Mongo
+{
+    public class MongoDbCollectionChecker
```
I understand the check but not sure I understand the purpose of the check. The wireup stops jobs from running which sounds efficient but what about QA web and employer web?

Previously we were checking the collection everytime an interaction was going to be made which I agree looks excessive. 

The reason for it was because `applicationReviews` collection was created in PROD without it being declared in the Cosmos ARM config template so it didn't have the correct setup. The previous check was to ensure no V2 component application would create the collection automatically.

sachpatel replied:
Yes please

chrisjensenuk replied:
👍

chrisjensenuk commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/Client/IEmployerVacancyClient.cs:
```diff
@@ -31,7 +31,7 @@ public interface IEmployerVacancyClient
         Task SetupEmployerAsync(string employerAccountId);
         Task<List<string>> GetCandidateSkillsAsync();
-        Task<Quals.Qualifications> GetCandidateQualificationsAsync();
+        Task<IList<string>> GetCandidateQualificationsAsync();
```
👍

chrisjensenuk commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/Mongo/MongoDbCollectionBase.cs:
```diff
@@ -24,21 +24,22 @@ protected MongoDbCollectionBase(ILogger logger, string dbName, string collection
             var settings = MongoClientSettings.FromUrl(new MongoUrl(_config.ConnectionString));
             settings.SslSettings = new SslSettings { EnabledSslProtocols = SslProtocols.Tls12 };
 
+#if DEBUG
```
I was unsure how best to implement this as its only useful when debugging so I didn't want to make it anymore complicated than it needs to be.  I may remove entirely and just add it as a code snippet in Confluence then we can just drop it it when we need to.  Would this be better?  Happy to hear other suggestions.

leewadhams replied:
I would probably agree with @sachpatel on this one in terms of passing in the IHostingEnvironment as this is what we use everywhere else to do something different only for local dev. We could also leave the line uncommented and add an exclude filter to the logging config if we don't want it to muddy our local logs.

sachpatel replied:
Yeah I think you having added this is good Chris and is a nice to have for local development.

chrisjensenuk replied:
I'll implement @leewadhams suggestion and add an exclude filter.

chrisjensenuk commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/ApprenticeshipProgrammes/ApprenticeshipProgrammeProvider.cs:
```diff
@@ -9,10 +11,14 @@ namespace Esfa.Recruit.Vacancies.Client.Infrastructure.ReferenceData.Apprentices
+        private readonly ICache _cache;
 
-        public ApprenticeshipProgrammeProvider(IReferenceDataReader queryStoreReader)
+        private DateTime CacheAbsoluteExpiryTime => DateTime.UtcNow.Date.AddDays(1).AddHours(4).AddMinutes(15);
```
Can do 6am

dev4valtech replied:
Since this is done only once a day, there should be a job in place to kill the cache if we ever need it.

chrisjensenuk replied:
@dev4valtech For now we can stop and start the services/web apps to clear the caches.  I'll add an item to the backlog to come up with less extreme way!

chrisjensenuk commented on src/Shared/Recruit.Vacancies.Client/Ioc/ServiceCollectionExtensions.cs:
```diff
@@ -65,7 +67,7 @@ public static void AddRecruitStorageClient(this IServiceCollection services, ICo
 
             services.AddTransient<IConfigurationReader, ConfigurationReader>();
-            services.AddTransient<QaRulesConfiguration>(x => 
+            services.AddSingleton<QaRulesConfiguration>(x => 
```
👍

chrisjensenuk replied:
Will revert

chrisjensenuk commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayProvider.cs:
```diff
@@ -11,19 +12,28 @@ namespace Esfa.Recruit.Vacancies.Client.Infrastructure.Services
+        private readonly ICache _cache;
 
-        public BankHolidayProvider(IReferenceDataReader referenceDataReader)
+        private DateTime CacheAbsoluteExpiryTime => DateTime.UtcNow.Date.AddDays(1);
```
I'll move this Times.cs into Client and reuse that

chrisjensenuk replied:
I'll delete Times.cs entirely and I'll use ITimeProvider instead so we have a single place for date manipulations.

chrisjensenuk commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayUpdateService.cs:
```diff
@@ -29,6 +29,9 @@ public async Task UpdateBankHolidaysAsync()
+                throw new Exception($"Error getting list of bank holidays from url:{_config.Url}. Error:{response.ErrorMessage}");
+
             if (!response.Data.EnglandAndWales.Events.Any())
                 throw new Exception($"Expected a list of bank holidays from url:{_config.Url}");
```
👍

dev4valtech commented:
I suggest that in future we do smaller PR's it would be so much easy to review and pull in changes that are approved.

leewadhams commented on src/Jobs/Recruit.Vacancies.Jobs/Program.cs:
```diff
@@ -76,8 +77,16 @@ public static async Task Main(string[] args)
             var host = builder.Build();
             using (host)
             {
+                CheckInfrastructure(host.Services);
```
I don't think we'll get any logging of the exception would we?

chrisjensenuk replied:
I'll put back the "The Job has met with a horrible end!!". I miss it.

leewadhams commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/ReferenceData/BankHolidays/BankHolidayProvider.cs:
```diff
@@ -11,19 +12,28 @@ namespace Esfa.Recruit.Vacancies.Client.Infrastructure.Services
-            var bankHolidayReferenceData = await _referenceDataReader.GetReferenceData<BankHolidays>();
+            return await _cache.CacheAsideAsync(CacheKeys.BankHolidays,
+                CacheAbsoluteExpiryTime,
+                async () =>
```
On a couple of the providers I think we have some inconsistency in terms of the null checks and logging.

dev4valtech commented on src/Shared/Recruit.Vacancies.Client/Infrastructure/Repositories/MongoDbVacancyReviewRepository.cs:
```diff
@@ -79,19 +83,21 @@ public async Task<List<VacancyReview>> GetByStatusAsync(ReviewStatus status)
 
-            var filter = filterBuilder.Eq(r => r.Status, ReviewStatus.PendingReview)
-                         | filterBuilder.Eq(r => r.Status, ReviewStatus.UnderReview);
+            var filter = filterBuilder.Eq(Status, ReviewStatus.PendingReview.ToString())
```
Could use a base class and put a generic filter `where T: BaseClass` and carry on using Lambda expression here.

chrisjensenuk replied:
At the moment it works with any type you throw at it. I don't think making it tighter coupled is worth the lambda expression payoff.

dev4valtech replied:
I think it is a matter of setting expectations rather than tight coupling.

chrisjensenuk replied:
There is no reason why the Type you are projecting to should contain the properties your are filtering on. In this case were projecting to `VacancyReviewSummary` which doesn't have `Status` as a property.

chrisjensenuk commented on src/Shared/Recruit.Vacancies.Client/Domain/Entities/VacancyReviewSummary.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Esfa.Recruit.Vacancies.Client.Domain.Entities
+{
+    public class VacancyReviewSummary
```
Its a different sort of "projection" as its coming from the Repo and not the QueryStore.  We have a few of these under the Entities folder maybe we should review and split into different folders. `Entities` and `ReadEntities` (or something)  to follow the interface convention on the repos. 

I won't change now but will add a note to tech debt.

chrisjensenuk commented:
Sorry all. I've added loads more changes.

- `IHostingEnvironment` only exists in `AspNetCore` so this won't work the for Client Library or Jobs.  I've moved the existing environment logic we use in Jobs into a new class `RecruitEnvironment` and both the Client library and Jobs now use this
- I've add a `LoggingBuilderExtensions` class that explicitly does not log mongo commands.  If you do want to log the commands change the filter to `LogLevel.Trace`
- Deleted `Times.cs` and using `ITimeProvider` instead
- Now checking that collections exist when Employer & QA start-up. Have re-added the "The Job has met with a horrible end!!" logging to Jobs.

leewadhams commented on src/Jobs/Recruit.Vacancies.Jobs/Program.cs:
```diff
@@ -17,65 +18,84 @@ class Program
-                    .AddJsonFile($"appSettings.{RecruitEnvironment.EnvironmentName}.json", true)
-                    .AddEnvironmentVariables();
+                    logger = ((ILoggerFactory)host.Services.GetService(typeof(ILoggerFactory)))
+                        .CreateLogger("Program");
```
nameof ?

chrisjensenuk replied:
👍
